homeassistant/components/switch/tplink.py

```diff
@@ -7,58 +7,68 @@
 import logging
 import time
 
-import voluptuous as vol
-
 from homeassistant.components.switch import (
-    SwitchDevice, PLATFORM_SCHEMA, ATTR_CURRENT_POWER_W, ATTR_TODAY_ENERGY_KWH)
-from homeassistant.const import (CONF_HOST, CONF_NAME, ATTR_VOLTAGE)
-import homeassistant.helpers.config_validation as cv
+    SwitchDevice, ATTR_CURRENT_POWER_W, ATTR_TODAY_ENERGY_KWH)
+from homeassistant.components.tplink import DOMAIN as TPLINK_DOMAIN
+from homeassistant.const import ATTR_VOLTAGE
+import homeassistant.helpers.device_registry as dr
+
+DEPENDENCIES = ['tplink']
 
-REQUIREMENTS = ['pyHS100==0.3.3']
+PARALLEL_UPDATES = 0
 
 _LOGGER = logging.getLogger(__name__)
 
 ATTR_TOTAL_ENERGY_KWH = 'total_energy_kwh'
 ATTR_CURRENT_A = 'current_a'
 
-CONF_LEDS = 'enable_leds'
-
-DEFAULT_NAME = 'TP-Link Switch'
 
-PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
-    vol.Required(CONF_HOST): cv.string,
-    vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
-    vol.Optional(CONF_LEDS): cv.boolean,
-})
+async def async_setup_entry(hass, config_entry, async_add_devices):
+    """Set up discovered switches."""
+    devs = []
+    for dev in hass.data[TPLINK_DOMAIN]['switch']:
+        devs.append(SmartPlugSwitch(dev))
 
-
-def setup_platform(hass, config, add_entities, discovery_info=None):
-    """Set up the TPLink switch platform."""
-    from pyHS100 import SmartPlug
-    host = config.get(CONF_HOST)
-    name = config.get(CONF_NAME)
-    leds_on = config.get(CONF_LEDS)
-
-    add_entities([SmartPlugSwitch(SmartPlug(host), name, leds_on)], True)
+    async_add_devices(devs, True)
 
 
 class SmartPlugSwitch(SwitchDevice):
     """Representation of a TPLink Smart Plug switch."""
 
-    def __init__(self, smartplug, name, leds_on):
+    def __init__(self, smartplug):
         """Initialize the switch."""
         self.smartplug = smartplug
-        self._name = name
-        self._leds_on = leds_on
+        self._sysinfo = None
         self._state = None
-        self._available = True
+        self._available = False
         # Set up emeter cache
         self._emeter_params = {}
 
+    @property
+    def unique_id(self):
+        """Return a unique ID."""
+        return self._sysinfo["mac"]
+
     @property
     def name(self):
-        """Return the name of the Smart Plug, if any."""
-        return self._name
+        """Return the name of the Smart Plug."""
+        return self._sysinfo["alias"]
+
+    @property
+    def device_info(self):
+        """Return information about the device."""
+        return {
+            'identifiers': {
+                (TPLINK_DOMAIN, self.unique_id)
```

Comment:
Remove this. It's not a TP Link provided identifier, it's a Mac address and it's already provided as a connections.

Reply:
Ok, I removed it. There is also two identifiers which could be used here if wanted: `deviceId` and `hwId`, but that probably doesn't change anything if the mac is already there?